### PR TITLE
Add const annotations to CommonJS module exports

### PR DESCRIPTION
**Pull request by ChadKillingsworth**
To help get CommonJS to parity with ES6 modules, this changes the CommonJS module rewriting to add `@const` annotations if we can reasonably be certain that the export property was assigned exactly once.

### CODE REVIEW

**Comment by ChadKillingsworth** [comment]
Merge conflicts resolved. Can this be reviewed now?

File: src/com/google/javascript/jscomp/ProcessCommonJSModules.java [hunk 1]
```diff
@@ -250,10 +255,48 @@ public void visit(NodeTraversal t, Node n, Node parent) {
...
+    private void maybeAddReferenceCount(Node n) {
+      Node refParent = getBaseQualifiedNameNode(n);
+
+      if (refParent == null) {
+        return;
+      }
+
+      String qName = refParent.getQualifiedName();
+      if (qName.startsWith("module.exports.")) {
+        qName = qName.substring(15);
```
Comment (MatrixFrog) [inline src/com/google/javascript/jscomp/ProcessCommonJSModules.java]: nit: use `"module.exports.".length` so it's clear where that number comes from (likewise 2 lines down)

Comment (ChadKillingsworth) [inline src/com/google/javascript/jscomp/ProcessCommonJSModules.java]: Done

File: src/com/google/javascript/jscomp/ProcessCommonJSModules.java [hunk 2]
```diff
@@ -377,57 +420,36 @@ private void processExports(Node script, String moduleName) {
...
-          }
+          addConstToObjLitKeys(rhsValue);
         }
 
         Node assign = ref.getParent();
-        Node exprResult = assign.getParent();
-        script.replaceChild(exprResult, newExprResult);
+        assign.replaceChild(ref, newName);
+        JSDocInfoBuilder builder = new JSDocInfoBuilder(true);
+        builder.recordConstancy();
```
Comment (MatrixFrog) [inline src/com/google/javascript/jscomp/ProcessCommonJSModules.java]: I would expect to see these `@const` annotations in the tests, but I guess we don't have the compareJsDoc option enabled for this test? We should enable it but I don't think we need to block this PR on it.

File: src/com/google/javascript/jscomp/ProcessCommonJSModules.java [hunk 3]
```diff
@@ -179,6 +182,7 @@ public boolean shouldTraverse(NodeTraversal nodeTraversal, Node n, Node parent)
     private int scriptNodeCount = 0;
     private List<Node> moduleExportRefs = new ArrayList<>();
     private List<Node> exportRefs = new ArrayList<>();
+    Map<String, Integer> propertyExportRefCount = new HashMap<>();
```
Comment (MatrixFrog) [inline src/com/google/javascript/jscomp/ProcessCommonJSModules.java]: You can often replace a `Map<String, Integer>` with a `Multiset<Key>` and I think this is a case where that would work well, but I'm not sure. Take a look and if it switching to a Multiset is a lot of work then nevermind but if it turns out to be easy let's do it.

http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/collect/Multiset.html

**Comment by MatrixFrog** [comment]
otherwise lgtm

File: src/com/google/javascript/jscomp/ProcessCommonJSModules.java (hunk 3 above)
Comment (ChadKillingsworth) [inline src/com/google/javascript/jscomp/ProcessCommonJSModules.java]: Simple change - I just don't use Guava much.

File: src/com/google/javascript/jscomp/ProcessCommonJSModules.java (hunk 2 above)
Comment (ChadKillingsworth) [inline src/com/google/javascript/jscomp/ProcessCommonJSModules.java]: It's definitely disabled - and enabling looks to be a non-trivial amount of work.

**Comment by ChadKillingsworth** [comment]
Toggled to force travis to run.

File: src/com/google/javascript/jscomp/ProcessCommonJSModules.java (hunk 2 above)
Comment (Dominator008) [inline src/com/google/javascript/jscomp/ProcessCommonJSModules.java]: Maybe add a test to ProcessCommonJSModulesTest after #1395 is merged?

Comment (ChadKillingsworth) [inline src/com/google/javascript/jscomp/ProcessCommonJSModules.java]: I actually think the existing tests will have to be updated and then it should be covered. I could pull #1395 into here, but they are based off different master commits so it makes a mess of the tree.

Comment (Dominator008) [inline src/com/google/javascript/jscomp/ProcessCommonJSModules.java]: We can merge this first :)

**Pull request merged**